client/spawn/spawn.go

```diff
@@ -0,0 +1,309 @@
+		ExitStatusFile: s.StateFile,
+	}
+
+	if s.Logs != nil {
+		config.StdoutFile = s.Logs.Stdout
+		config.StdinFile = s.Logs.Stdin
+		config.StderrFile = s.Logs.Stderr
+	}
+
+	var buffer bytes.Buffer
+	enc := json.NewEncoder(&buffer)
+	if err := enc.Encode(config); err != nil {
+		return "", fmt.Errorf("Failed to serialize configuration: %v", err)
+	}
+
+	return strconv.Quote(buffer.String()), nil
+}
+
+// sendStartCommand sends the necessary command to the spawn-daemon to have it
+// start the user process.
+func (s *Spawner) sendStartCommand(w io.Writer) error {
+	enc := json.NewEncoder(w)
+	if err := enc.Encode(true); err != nil {
+		return fmt.Errorf("Failed to serialize start command: %v", err)
+	}
+
+	return nil
+}
+
+// sendAbortCommand sends the necessary command to the spawn-daemon to have it
+// abort starting the user process. This should be invoked if the spawn-daemon
+// could not be isolated into a cgroup.
+func (s *Spawner) sendAbortCommand(w io.Writer) error {
+	enc := json.NewEncoder(w)
+	if err := enc.Encode(false); err != nil {
+		return fmt.Errorf("Failed to serialize abort command: %v", err)
+	}
+
+	return nil
+}
+
+// Wait returns the exit code of the user process or an error if the wait
+// failed.
+func (s *Spawner) Wait() (int, error) {
+	if os.Getpid() == s.SpawnPpid {
+		return s.waitAsParent()
+	}
+
+	return s.waitOnStatusFile()
+}
+
+// waitAsParent waits on the process if the current process was the spawner.
+func (s *Spawner) waitAsParent() (int, error) {
+	if s.SpawnPpid != os.Getpid() {
+		return -1, fmt.Errorf("not the parent. Spawner parent is %v; current pid is %v", s.SpawnPpid, os.Getpid())
+	}
+
+	// Try to reattach to the spawn.
+	if s.spawn == nil {
+		// If it can't be reattached, it means the spawn process has exited so
+		// we should just read its exit file.
+		var err error
+		if s.spawn, err = os.FindProcess(s.SpawnPid); err != nil {
+			return s.waitOnStatusFile()
+		}
+	}
+
+	if _, err := s.spawn.Wait(); err != nil {
+		return -1, err
+	}
+
+	return s.waitOnStatusFile()
+}
+
+// waitOnStatusFile uses OS level file watching APIs to wait on the status file
+// and returns the exit code and possibly an error.
+func (s *Spawner) waitOnStatusFile() (int, error) {
+	// Set up a watcher for the exit status file.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return -1, fmt.Errorf("Failed to create file watcher to read exit code: %v", err)
+	}
+
+	if err := watcher.Add(s.StateFile); err != nil {
+		return -1, fmt.Errorf("Failed to watch %v to read exit code: %v", s.StateFile, err)
+	}
+
+	// Stat to check if it is there to avoid a race condition.
+	stat, err := os.Stat(s.StateFile)
+	if err != nil {
+		return -1, fmt.Errorf("Failed to Stat exit status file %v: %v", s.StateFile, err)
+	}
+
+	// If there is data it means that the file has already been written.
+	if stat.Size() > 0 {
+		return s.readExitCode()
+	}
+
+	// Wait on watcher.
+	for {
+		select {
+		case event := <-watcher.Events:
```

Comment:
Are we watching for events to the file where spawn daemon is going to write the exit code for the process it starts using inotify and also checking whether the spawn daemon is alive every 5s at the same time?

If we are checking if spawn daemon is alive every 5s, then we could just read the file for the exit code when spwan daemon has exited.